sql/2020/11_Security/iframe_attribute_popular_hosts.sql

```diff
@@ -0,0 +1,38 @@
+#standardSQL
+# most common hostnames of iframes that have the allow or sandbox attribute
+CREATE TEMP FUNCTION hasPolicy(attr STRING, policy_type STRING)
+  RETURNS BOOL DETERMINISTIC
+  LANGUAGE js AS '''
+  const $ = JSON.parse(attr);
+  return $[policy_type] !== null;
+''';
+
+SELECT
+  client,
+  policy_type,
+  hostname,
+  SUM(COUNTIF(has_policy)) OVER (PARTITION BY client, policy_type) AS total,
+  COUNTIF(has_policy) AS freq,
+  COUNTIF(has_policy) / SUM(COUNTIF(has_policy)) OVER (PARTITION BY client, policy_type) AS pct
+FROM (
+  SELECT
+    client,
+    policy_type,
+    JSON_EXTRACT_SCALAR(iframeAttr, '$.hostname') AS hostname,
+    hasPolicy(iframeAttr, policy_type) AS has_policy
+  FROM (
+    SELECT
+      _TABLE_SUFFIX as client,
+      JSON_EXTRACT_ARRAY(JSON_EXTRACT_SCALAR(payload, '$._security'), "$.iframe-allow-sandbox") AS iframeAttrs
+    FROM
+      `httparchive.pages.2020_08_01_*`),
+    UNNEST(iframeAttrs) AS iframeAttr,
+    UNNEST(['allow', 'sandbox']) AS policy_type
+  )
+GROUP BY
+  client,
+  policy_type,
+  hostname
+ORDER BY
+  pct DESC
+LIMIT 100
```

Comment:
I'm not exactly sure how to do that; I get the following error when trying to add a `HAVING pct >` clause:

> Column pct contains an analytic function, which is not allowed in HAVING clause at [37:3]

Reply:
Something like this works, but probably it's possible to get something nicer...

```sql
#standardSQL
# most common hostnames of iframes that have the allow or sandbox attribute
CREATE TEMP FUNCTION hasPolicy(attr STRING, policy_type STRING)
  RETURNS BOOL DETERMINISTIC
  LANGUAGE js AS '''
  const $ = JSON.parse(attr);
  return $[policy_type] !== null;
''';

SELECT
  *
FROM (
  SELECT
    client,
    policy_type,
    hostname,
    SUM(COUNTIF(has_policy)) OVER (PARTITION BY client, policy_type) AS total_iframes,
    COUNTIF(has_policy) AS freq,
    COUNTIF(has_policy) / SUM(COUNTIF(has_policy)) OVER (PARTITION BY client, policy_type) AS pct
  FROM (
    SELECT
      client,
      policy_type,
      JSON_EXTRACT_SCALAR(iframeAttr, '$.hostname') AS hostname,
      hasPolicy(iframeAttr, policy_type) AS has_policy
    FROM (
      SELECT
        _TABLE_SUFFIX as client,
        JSON_EXTRACT_ARRAY(JSON_EXTRACT_SCALAR(payload, '$._security'), "$.iframe-allow-sandbox") AS iframeAttrs
      FROM
        `httparchive.sample_data.pages_*`),
      UNNEST(iframeAttrs) AS iframeAttr,
      UNNEST(['allow', 'sandbox']) AS policy_type
    )
  GROUP BY
    client,
    policy_type,
    hostname
)
WHERE
  pct > 0.001
ORDER BY
  pct DESC
```

Reply:
Saw same thing in another PR I reviewed. Wasn't aware complex `pct` calcs couldn't be included.

Suggest maybe limiting on `freq` then instead?

Maybe LIMIT is fine, but just feels like potentially could be cutting off a load of queries that just so happen to fall out of this. Where as limiting on `freq` or `pct` at least means you are looking at values over a certain threshold rather than just top 100 values. Do you get me?

Reply:
How about this one? It just calculates `total_iframes` differently:

```sql
#standardSQL
# most common hostnames of iframes that have the allow or sandbox attribute
CREATE TEMP FUNCTION hasPolicy(attr STRING, policy_type STRING)
  RETURNS BOOL DETERMINISTIC
  LANGUAGE js AS '''
  const $ = JSON.parse(attr);
  return $[policy_type] !== null;
''';

SELECT
  client,
  policy_type,
  hostname,
  total_iframes,
  COUNTIF(has_policy) AS freq,
  COUNTIF(has_policy) / total_iframes AS pct
FROM (
  SELECT
    client,
    policy_type,
    JSON_EXTRACT_SCALAR(iframeAttr, '$.hostname') AS hostname,
    hasPolicy(iframeAttr, policy_type) AS has_policy
  FROM (
    SELECT
      _TABLE_SUFFIX as client,
      JSON_EXTRACT_ARRAY(JSON_EXTRACT_SCALAR(payload, '$._security'), "$.iframe-allow-sandbox") AS iframeAttrs
    FROM
      `httparchive.sample_data.pages_*`),
    UNNEST(iframeAttrs) AS iframeAttr,
    UNNEST(['allow', 'sandbox']) AS policy_type
)
JOIN (
  SELECT
    _TABLE_SUFFIX AS client,
    SUM(ARRAY_LENGTH(JSON_EXTRACT_ARRAY(JSON_EXTRACT_SCALAR(payload, '$._security'), "$.iframe-allow-sandbox"))) AS total_iframes
  FROM
    `httparchive.sample_data.pages_*`
  GROUP BY
    client)
USING
  (client)
GROUP BY
  client,
  total_iframes,
  policy_type,
  hostname
HAVING
  pct > 0.001
ORDER BY
  client,
  pct DESC
```

Reply:
Yeah that looks good to me. Obviously change from sample_data when submitting in this PR.